### PR TITLE
OPENTOK-48681: Avoid crash on cast

### PR DESCRIPTION
**Pull request by goncalocostamendes**
What is this PR doing?

Add range to audio playout delay, in case of incorrect inputs.

Ticket: https://jira.vonage.com/browse/OPENTOK-48681

### CODE REVIEW

File: Custom-Audio-Driver/Custom-Audio-Driver/DefaultAudioDevice.swift [hunk 1]
```diff
@@ -630,20 +632,33 @@ func updatePlayoutDelay(withAudioDevice audioDevice: DefaultAudioDevice) {
...
+        tempPlayoutDelay += UInt32(interval * 1000000)
         // HW buffer duration
         let ioInterval = session.ioBufferDuration
-        audioDevice.playoutDelay += UInt32(ioInterval * 1000000)
-        audioDevice.playoutDelay += UInt32(audioDevice.playoutAudioUnitPropertyLatency * 1000000)
+        tempPlayoutDelay += UInt32(ioInterval * 1000000)
+        tempPlayoutDelay += UInt32(audioDevice.playoutAudioUnitPropertyLatency * 1000000)
         // To ms
-        audioDevice.playoutDelay = (audioDevice.playoutDelay - 500) / 1000
+        if( tempPlayoutDelay >= 500 ) {
```
Comment (jvalli) [inline Custom-Audio-Driver/Custom-Audio-Driver/DefaultAudioDevice.swift:646]: Why do we use a StandardPlayoutDelay now?

**Comment by IGitGotIt** [comment]
The original JIRA issue was a type cast crash , so why are we adding  range check? What happened (before) if we go over the range ?

File: Custom-Audio-Driver/Custom-Audio-Driver/DefaultAudioDevice.swift [hunk 2]
```diff
@@ -630,20 +632,33 @@ func updatePlayoutDelay(withAudioDevice audioDevice: DefaultAudioDevice) {
...
     if audioDevice.playoutDelayMeasurementCounter >= 100 {
         // Update HW and OS delay every second, unlikely to change
-        audioDevice.playoutDelay = 0
+        var tempPlayoutDelay: UInt32 = 0
         let session = AVAudioSession.sharedInstance()
         
         // HW output latency
         let interval = session.outputLatency
-        audioDevice.playoutDelay += UInt32(interval * 1000000)
+        tempPlayoutDelay += UInt32(interval * 1000000)
```
[the remark this answers is not in the document]
Comment (goncalocostamendes) [reply Custom-Audio-Driver/Custom-Audio-Driver/DefaultAudioDevice.swift:640]: of course. Will do that

File: Custom-Audio-Driver/Custom-Audio-Driver/DefaultAudioDevice.swift (hunk 1 above)
Comment (goncalocostamendes) [reply Custom-Audio-Driver/Custom-Audio-Driver/DefaultAudioDevice.swift:646]: In the scenario that tempPlayoutDelay is lower than 500, to avoid having a negative value which will crash the system, we assign the default playoutDelay value (15). 

This value comes from my tests. I have tested several scenarios, including default audio driver embedded in multi-party session and the value audioDevice.playoutDelay is assign is always 15

What would you suggest? Skip and let it calculate in the next iteration?

**Comment by IGitGotIt** [comment]
> > The original JIRA issue was a type cast crash , so why are we adding range check? What happened (before) if we go over the range ?
> 
> @IGitGotIt When programming in swift, if the value goes over the variable size the program crashes. So there is an implicit range from 0 to 65535.
> 
> We do not have any traces of the exact scenarion when the error occurs, so it looks that maybe we get some [mySession outputLatency] nonsense value in some situations, which goes over the max value. Maybe when switching mute state.
> 
> By definition, playout delay is the total amount of time experienced by an audio packet from the time instant it is generated at the source and the time instant it is played out at the destination. So 65 seconds is a extremely high value. That being said, we can shorten this value. My idea was 1s, but we can increase it to something else.
> 
> FYI, I have tested several scenarios, including default audio driver embedded in multi-party session and the value audioDevice.playoutDelay is always assigned to 15. Of course, it should vary a bi

> > The original JIRA issue was a type cast crash , so why are we adding range check? What happened (before) if we go over the range ?
> 
> @IGitGotIt When programming in swift, if the value goes over the variable size the program crashes. So there is an implicit range from 0 to 65535.
> 
> We do not have any traces of the exact scenarion when the error occurs, so it looks that maybe we get some [mySession outputLatency] nonsense value in some situations, which goes over the max value. Maybe when switching mute state.
> 
> By definition, playout delay is the total amount of time experienced by an audio packet from the time instant it is generated at the source and the time instant it is played out at the destination. So 65 seconds is a extremely high value. That being said, we can shorten this value. My idea was 1s, but we can increase it to something else.
> 
> FYI, I have tested several scenarios, including default audio driver embedded in multi-party session and the value audioDevice.playoutDelay is always assigned to 15. Of course, it should vary a bit

First what happens if the playout delay is 65 or 1000 seconds? Does webrtc choke? or it self -corrects? or does nothing? What is the audio output in such case - lagging , jerky or ok? 

Also let **assume** webrtc self corrects or bypasses when values are out of line. Do we know the range of those values? What is the guarantee that your sane upper range value will kick in that self correction/bypass or not? It may never self correct and problem will worsen instead of getting fixed. In fact this is the concern the partner had in the original ticket, which we never addressed specifically.   This are just my assumptions w/o looking at webrtc code, so please test with some sample values and/or webrtc code check.

Now if after all this investigation you feel we need to do your change,  you could just keep the old code and add the following (pseudo-almost correct)code:
Basically control the logic where you set playoutDelay in one place for every assignment. 

```swift
    var _playoutDelay: UInt32 = 15 //some default
    var playoutDelay: UInt32 {
            get {
                return _playoutDelay
            }
            set(newValue) {
                 _playoutDelay = min(newValue,1000)
                if newValue > 1000 {
                    playoutDelayMeasurementCounter = 100
                }
            }
    }
```
In fact you can add the  whole `updatePlayoutDelay` in the set method above, if you want...

**Comment by goncalocostamendes** [comment]
@IGitGotIt I added the results of the tests to the jira ticekt https://jira.vonage.com/browse/OPENTOK-48681

**Comment by IGitGotIt** [comment]
> @IGitGotIt I added the results of the tests to the jira ticekt https://jira.vonage.com/browse/OPENTOK-48681. Briefly, with a 65000ms playout delay, the loss of audio packets increases considerable.

Thanks @goncalocostamendes .  
Here is a good article on what playout delay does - https://flylib.com/books/en/4.245.1.52/1/

The delay typically decides how long the frame buffer holds onto the data and thus the longer it is , the more incoming (new) frames will be dropped. Now here I think a 65 seconds or even 1 seconds will behave "almost" the same since what we are providing is an "estimatedCaptureDelay" and [RTP](https://chromium.googlesource.com/external/webrtc/+/branch-heads/4515/audio/channel_receive.cc#1020) will hopefully take over if the value we provides goes out of line. (a. guess). Maybe you can try out this experiment. If my assumption is correct.. then...

So if the delay value recovers by itself having a value of 65 or 1 will not matter. You can say let's do 150 ms or 300 ms. That will make sense. But if it recovers itself quickly in the next second (the value of 100)  why over-engineer it.

I would be ok leaving the code as it is and communicate the findings to the user. Even if we change the delay value, it is sort of a recommendation to RTP layer in extreme case. Also we providing a value which is different from what the Audio Unit is telling us now, does not mean Audio Unit will adjust. It will do what it wants to do. It is just a value for the RTP layer.

Let say we make a mistake programmatically/something messed up,  in giving a max value for estimatedCaptureDelay then we can just add a `return min(playoutDelay,300)` (ms)  like statement to https://github.com/opentok/opentok-ios-sdk-samples-swift/blob/main/Custom-Audio-Driver/Custom-Audio-Driver/DefaultAudioDevice.swift#L295.

Anyway try the experiment with 1 second vs 65 seconds vs 300 ms vs 100 ms and the "always/correct"  present value of 15 ms from your observations.

As a side note a value of 0 is also good..let RTP decide wit its internal default.. (maybe try it out) 

Thanks

**Comment by goncalocostamendes** [comment]
@IGitGotIt this online book is great!! I was looking for something like that.

I have on the PR, on top of the range, forcing that when a "wrong" value is set to the playout delay, instead of waiting for the next calculation (each one 1s), it will do it right on the next iteration (with the assumption that a correct value will be received in one of the next iterations). This being an assumption, maybe having only a validation between the value calculated and the max will be enough, and reduces the calls to several functions.

I found on the book:

_The trade-off in playout buffer operation is between fidelity and delay: An application must decide the maximum playout delay it can accept, and this in turn determines the fraction of packets that arrive in time to be played out. A system designed for interactive usefor example, video conferencing or telephonymust try to keep the playout delay as small as possible because it cannot afford the latency incurred by the buffering. Studies of human perception point to a limit in round-trip time of about 300 milliseconds as the maximum tolerable for interactive use; this limit implies an end-to-end delay of only 150 milliseconds including network transit time and buffering delay if the system is symmetric. However, a noninteractive system, such as streaming video, television, or radio, may allow the playout buffer to grow up to several seconds, thereby enabling noninteractive systems to handle variation in packet arrival times better._

Since the 150ms was reached by studies on human perception, I believe we should stick with it. What do you think?

**Review by sergioturgil-vonage — APPROVED** [review]
All good but i think it would be safer to use an integer primitive type for the constant

File: Custom-Audio-Driver/Custom-Audio-Driver/DefaultAudioDevice.swift [hunk 3]
```diff
@@ -20,6 +20,9 @@ class DefaultAudioDevice: NSObject {
     static let kAudioDeviceHeadset = "AudioSessionManagerDevice_Headset"
     static let kAudioDeviceBluetooth = "AudioSessionManagerDevice_Bluetooth"
     static let kAudioDeviceSpeaker = "AudioSessionManagerDevice_Speaker"
+    static let kStandardPlayoutDelay: UInt8 = 15
+    static let kToMicroSecond: Double = 1000000
```
[the remark this answers is not in the document]
Comment (goncalocostamendes) [reply Custom-Audio-Driver/Custom-Audio-Driver/DefaultAudioDevice.swift:24]: I made it a double to reduce casting, because it is always multiplicated by a double variable. Changing it to a uint32 and cast it to double every time will keep the problem. We can instead cast the double to int, but won't that also raise the same issue? And in this case, the variable is much smaller, causing an higher inaccuracy. 

Also, the result will always be divided by 1000 afterwards, already with both being an int. What do you think is the best approach?

File: Custom-Audio-Driver/Custom-Audio-Driver/DefaultAudioDevice.swift [hunk 4]
```diff
@@ -655,11 +659,11 @@ func updateRecordingDelay(withAudioDevice audioDevice: DefaultAudioDevice) {
...
         let interval = session.inputLatency
         
-        audioDevice.recordingDelayHWAndOS += UInt32(interval * 1000000)
+        audioDevice.recordingDelayHWAndOS += UInt32(interval * DefaultAudioDevice.kToMicroSecond)
         let ioInterval = session.ioBufferDuration
         
-        audioDevice.recordingDelayHWAndOS += UInt32(ioInterval * 1000000)
-        audioDevice.recordingDelayHWAndOS += UInt32(audioDevice.recordingAudioUnitPropertyLatency * 1000000)
+        audioDevice.recordingDelayHWAndOS += UInt32(ioInterval * DefaultAudioDevice.kToMicroSecond)
+        audioDevice.recordingDelayHWAndOS += UInt32(audioDevice.recordingAudioUnitPropertyLatency * DefaultAudioDevice.kToMicroSecond)
```
Comment (IGitGotIt) [inline Custom-Audio-Driver/Custom-Audio-Driver/DefaultAudioDevice.swift:666]: Can't we add all of them in one line like  " kToMicroSecond * (a+ b +c )"

File: Custom-Audio-Driver/Custom-Audio-Driver/DefaultAudioDevice.swift [hunk 5]
```diff
@@ -293,7 +296,7 @@ extension DefaultAudioDevice: OTAudioDevice {
         return recording
     }
     func estimatedRenderDelay() -> UInt16 {
-        return UInt16(playoutDelay)
+        return UInt16(min(self.playoutDelay, UInt32(DefaultAudioDevice.kMaxPlayoutDelay)))
     }
     func estimatedCaptureDelay() -> UInt16 {
         return UInt16(recordingDelay)
```
Comment (IGitGotIt) [inline Custom-Audio-Driver/Custom-Audio-Driver/DefaultAudioDevice.swift:302]: no UInt16 casting needed here??

File: Custom-Audio-Driver/Custom-Audio-Driver/DefaultAudioDevice.swift [hunk 6]
```diff
@@ -655,11 +659,11 @@ func updateRecordingDelay(withAudioDevice audioDevice: DefaultAudioDevice) {
         let session = AVAudioSession.sharedInstance()
         let interval = session.inputLatency
         
-        audioDevice.recordingDelayHWAndOS += UInt32(interval * 1000000)
+        audioDevice.recordingDelayHWAndOS += UInt32(interval * DefaultAudioDevice.kToMicroSecond)
         let ioInterval = session.ioBufferDuration
         
-        audioDevice.recordingDelayHWAndOS += UInt32(ioInterval * 1000000)
-        audioDevice.recordingDelayHWAndOS += UInt32(audioDevice.recordingAudioUnitPropertyLatency * 1000000)
+        audioDevice.recordingDelayHWAndOS += UInt32(ioInterval * DefaultAudioDevice.kToMicroSecond)
```
Comment (IGitGotIt) [inline Custom-Audio-Driver/Custom-Audio-Driver/DefaultAudioDevice.swift:665]: can we call this "recordingDelay" in line with "playoutDelay" ?

File: Custom-Audio-Driver/Custom-Audio-Driver/DefaultAudioDevice.swift (hunk 2 above)
[the remark this answers is not in the document]
Comment (jvalli) [reply Custom-Audio-Driver/Custom-Audio-Driver/DefaultAudioDevice.swift]: 👍

File: Custom-Audio-Driver/Custom-Audio-Driver/DefaultAudioDevice.swift (hunk 1 above)
Comment (jvalli) [reply Custom-Audio-Driver/Custom-Audio-Driver/DefaultAudioDevice.swift]: Ok, seems that we can remove kStandardPlayoutDelay now.

File: Custom-Audio-Driver/Custom-Audio-Driver/DefaultAudioDevice.swift (hunk 4 above)
Comment (goncalocostamendes) [reply Custom-Audio-Driver/Custom-Audio-Driver/DefaultAudioDevice.swift:666]: In my opinion it would indeed make more sense, but in all other SDK samples and even in our native repo we have it like this

File: Custom-Audio-Driver/Custom-Audio-Driver/DefaultAudioDevice.swift (hunk 5 above)
Comment (goncalocostamendes) [reply Custom-Audio-Driver/Custom-Audio-Driver/DefaultAudioDevice.swift:302]: we need it because it is an uint32

File: Custom-Audio-Driver/Custom-Audio-Driver/DefaultAudioDevice.swift (hunk 6 above)
Comment (goncalocostamendes) [reply Custom-Audio-Driver/Custom-Audio-Driver/DefaultAudioDevice.swift:665]: They are different things. This particular change it was only regarding moving the 1000000 into a variable. Nothing else was changed here

Comment (IGitGotIt) [reply Custom-Audio-Driver/Custom-Audio-Driver/DefaultAudioDevice.swift:665]: I meant the var name

File: Custom-Audio-Driver/Custom-Audio-Driver/DefaultAudioDevice.swift (hunk 5 above)
Comment (IGitGotIt) [reply Custom-Audio-Driver/Custom-Audio-Driver/DefaultAudioDevice.swift:302]: hmm looks like I may have commented on old code.. UInt16  cast ok now
why not add a range here too?

File: Custom-Audio-Driver/Custom-Audio-Driver/DefaultAudioDevice.swift (hunk 4 above)
Comment (IGitGotIt) [reply Custom-Audio-Driver/Custom-Audio-Driver/DefaultAudioDevice.swift:666]: the legacy continues ....

File: Custom-Audio-Driver/Custom-Audio-Driver/DefaultAudioDevice.swift (hunk 5 above)
Comment (goncalocostamendes) [reply Custom-Audio-Driver/Custom-Audio-Driver/DefaultAudioDevice.swift:302]: I completely agree it should be validated, because the same issue we were facing with playoutDelay can also occur. I created https://jira.vonage.com/browse/OPENTOK-49302 for tech debt day

**Review by IGitGotIt — APPROVED** [review]
+1

**Review by jvalli — APPROVED** [review]
👍

**Comment by jvalli** [comment]
Do we know why the CI is failing?

**Comment by goncalocostamendes** [comment]
@jvalli yes, Custom-Video-Driver has a couple of compile errors. I created a ticket to solve that on tech-debt day https://jira.vonage.com/browse/OPENTOK-49279
